src/kernel/thread.c

```diff
@@ -91,6 +91,10 @@ void restart(tcb_t *target)
         cancelIPC(target);
 #ifdef CONFIG_KERNEL_MCS
         setThreadState(target, ThreadState_Restart);
+        if (target->tcbSchedContext != NULL) {
```

Comment:
> To implement sporadic servers, we need to ensure that the suspension of
a task cannot be used as a mechanism to amplify budget of a task by
granting that task access to effectively multiple periods worth of
replenishments within a single period.

This was already being achieved before this commit right?

> To align the implementation of SCs with the model of sporadic servers we
must delay available time until the release of a task. Within seL4, a
release would be any time where an SC changes from not being associated
with a Running, RunningVM, or Restart thread to one that is.

Does this mean that a new period can't start until a thread is released and any consumed budget won't be replenished until the new period starts.  Previously, each amount of consumed budget would be replenished one period after it was previously used whereas now it will also need to be released where being scheduled doesn't count as a release?

> Critically, replenishments should not be delayed at the point when an SC
becomes the current SC (as was the case prior to this commit). This has
the effect of enforcing a continuous, constant bandwidth which is a
restriction that is incompatible with standard scheduling logic.

I'm having trouble understanding this part of the message.  `refill_unblock_check` moves refills that are in the past to the present/future.  I assume that the commit is using "delayed" here to mean the same as made available but normally delayed means to move the rTime into the future such that it is greater than before and also greater than the present time.

Reply:
> This was already being achieved before this commit right?

This is the role of `refill_unblock_check` which was already doing this.

> Does this mean that a new period can't start until a thread is released and any consumed budget won't be replenished until the new period starts. Previously, each amount of consumed budget would be replenished one period after it was previously used whereas now it will also need to be released where being scheduled doesn't count as a release?

Each replenishment essentially acts as its own periodic task. That replenishment's period starts at its head time and subsequent periods are one later than the head time. If a thread blocks, all replenishments that would start before the time that thread unblocks are delayed and merged into a single replenishment with a period starting at the unblock time. Being scheduled still counts as a release or activation.

> I'm having trouble understanding this part of the message. refill_unblock_check moves refills that are in the past to the present/future. I assume that the commit is using "delayed" here to mean the same as made available but normally delayed means to move the rTime into the future such that it is greater than before and also greater than the present time.

This was referring to the fact that 'refill_unblock_check' was being called at the point where we switch to a task rather than when that task becomes schedulable. This causes a preempted thread to be scheduled as though it had blocked for the duration of the preemption which we do not want.

Reply:
> Being scheduled still counts as a release or activation.

Doesn't being scheduled mean being picked by the scheduler to be switched to? So when a higher priority thread blocks or suspends and the scheduler picks a lower priority thread, this doesn't count as an activation right?

Reply:
Right, I should say that a thread being dequeued from the release queue is also an activation, i.e. when a task god from having no time available for execution to any at all while not being blocked.